engine/src/main/java/org/terasology/internet/internal/InternetManagerImpl.java

```diff
@@ -26,32 +35,63 @@
 import java.security.AccessController;
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
-import java.util.Set;
-import java.util.TreeSet;
+import java.util.Collections;
+import java.util.Map;
 
 /**
  * Implements InternetManager.
  */
 public class InternetManagerImpl implements InternetManager {
+    /**
+     * The module this InternetManager belongs to.
+     */
+    //@Nonnull
     private final Name moduleId;
 
-    private final Set<String> allowedHosts;
+    /**
+     * The allowed hosts.
+     */
+    //@Nonnull
+    private final Map<String, TIntSet> allowedHosts;
 
     /**
      * Construct a new InternetManagerImpl with a shared "permanent" cache.
      *
      * @param moduleId The module ID.
+     * @param context The current context.
      */
-    public InternetManagerImpl(Name moduleId) {
+    public InternetManagerImpl(Name moduleId, Context context) {
         this.moduleId = moduleId;
-        allowedHosts = new TreeSet<>();
-        allowedHosts.add("irc.freenode.net");
+        Config theConfig = context.get(Config.class);
+        if (theConfig == null) {
+            throw new IllegalStateException();
+        }
+        // TODO check if this could miss config updates, and make thread safe.
+        InternetConfig.Hosts hosts = theConfig.getInternet().getTcpWhitelists().get(moduleId);
+        if (hosts == null) {
+            allowedHosts = Collections.emptyMap();
+            return;
+        }
+        allowedHosts = Maps.transformValues(hosts.getHosts(), (x) -> x != null ? TCollections.unmodifiableSet(x.getPorts()) : null);
     }
 
     @Override
     public TCPSocket openTCPConnection(String hostname, int port) throws IOException {
-        if (!allowedHosts.contains(hostname)) {
-            throw new IllegalArgumentException();
+        Map<String, TIntSet> allowedHosts = this.allowedHosts;
+        if (!(allowedHosts.containsKey(hostname) || allowedHosts.containsKey("*"))) {
+            // neither hostname nor wildcard in allowedHosts
+            // TODO maybe use a different exception type? (SandboxException or something?)
+            throw new IllegalArgumentException("Blocked hostname");
+        }
+        // TODO We have many options when both a wildcard entry and a specific entry exist. Which one should we pick?
+        // 1. Merge allowed ports.
+        // 2. Wildcard overrides specific hostname.
+        // 3. Specific hostname overrides wildcard. This option pleases me the most. (It also enables blacklists.)
```

Comment:
No, wildcards don't actually work (yet). `*.example.com` would only match a literal `*.example.com`. I only added `*` as an option so a dedicated "HTTP" module (not an engine subsystem) is a possibility.

But if you had `*` set to all ports allowed, and you added `example.com` with only port 80 allowed, then `example.com` would match first and override the `*` match - only port 80 would be valid for `example.com`.

If I do get around to implementing wildcards, then `*.example.com` will match both `foo.example.com` and `bar.example.com`. If `bar.example.com` exists on the whitelist, then `bar.example.com` would only match against it, and `foo.example.com` would still use the `*.example.com` entry.

This allows the "whitelist" to act as both a whitelist *and a blacklist*.